File: Data Structures & Algorithms/pow-x-n/submission-2.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if(n==0) return 1;
        if(x==0 || n==1) return x;

        int power = abs(n);
        double res = 1;
        while(power){
            if(power & 1){
                res = res * x;
            }
            x*=x;
            power = power >> 1;
        }

        return n>0 ? res : 1/res;
    }
};

File: Data Structures & Algorithms/triangle/submission-2.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        dp = {}

        def dfs(i, j):
            if i == len(triangle):
                return 0

            if j == len(triangle[i]):
                return 2 * 10**6

            if (i,j) in dp:
                return dp[(i,j)]

            res = 2 * 10**6
            for nxt in range(2):
                nxt_j = nxt + j
                res = min(res, triangle[i][j] + dfs(i+1, nxt_j))
            
            dp[(i,j)] = res
            return res
        
        return dfs(0, 0)


                
        